### PR TITLE
Split reconciliation logic to it's own service

### PR DESCRIPTION
**Pull request by toban**
- Splits out the reconciliation part of EditEndpoint to it's own service
- Adds some unit tests
- Currently never hitting ->die which is being refactored out.

Bug: T282243

### CODE REVIEW

**Comment by toban** [comment]
needs a rebase

File: includes/ServiceWiring.php [hunk 1]
```diff
@@ -12,4 +14,16 @@
...
 
+	'WikibaseReconcileEdit.ReconciliationService' => function ( MediaWikiServices $services ): ReconciliationService {
+		$repo = WikibaseRepo::getDefaultInstance();
+
+		return new ReconciliationService(
+			$repo->getEntityIdLookup(),
+			$repo->getEntityLookup(),
+			$repo->getEntityRevisionLookup(),
+			$repo->newIdGenerator(),
+			$services->get( 'WikibaseReconcileEdit.ExternalLinks' )
```
Comment (lucaswerkmeister) [inline includes/ServiceWiring.php]: We have a convenience method for that, `WikibaseReconcileEditServices::getExternalLinks( $services )`.

File: includes/ServiceWiring.php [hunk 2]
```diff
@@ -12,4 +14,16 @@
 		);
 	},
 
+	'WikibaseReconcileEdit.ReconciliationService' => function ( MediaWikiServices $services ): ReconciliationService {
```
Comment (lucaswerkmeister) [inline includes/ServiceWiring.php:38]: This new service should also have an accessor method in `WikibaseReconcileEditServices`.

File: tests/integration/MediaWiki/Api/EditEndpointTest.php [hunk 3]
```diff
@@ -46,13 +47,15 @@ private function newHandler() {
...
 		return new EditEndpoint(
 			$repo->newEditEntityFactory(),
-			$repo->getEntityIdLookup(),
-			$repo->getEntityLookup(),
-			$repo->getEntityRevisionLookup(),
-			$repo->newIdGenerator(),
 			$propertyDataTypeLookup,
-			new ExternalLinks(
-				LoadBalancerSingle::newFromConnection( $this->db )
+			new ReconciliationService(
```
Comment (lucaswerkmeister) [inline tests/integration/MediaWiki/Api/EditEndpointTest.php]: We could probably replace this with `WikibaseReconcileEditServices::getReconciliationService()`, once that method exists – since we’re not injecting any special services in here. (That would replace the `LoadBalancerSingle` with whatever real load balancer is in the service container, but I assume that won’t make a difference in practice.)

File: tests/unit/MediaWiki/ReconciliationServiceTest.php [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+	public function testBasicConstruction() {
+		$service = new ReconciliationService(
+			$this->createMock( EntityIdLookup::class ),
+			$this->createMock( EntityLookup::class ),
+			$this->createMock( EntityRevisionLookup::class ),
+			WikibaseRepo::getDefaultInstance()->newIdGenerator(),
+			$this->createMock( ExternalLinks::class ),
+		);
+
+		$this->assertInstanceOf( ReconciliationService::class, $service );
```
Comment (lucaswerkmeister) [inline tests/unit/MediaWiki/ReconciliationServiceTest.php]: This assertion is guaranteed to succeed, since we literally created the object with `new ReconciliationService()`… I’m not sure what the point of this test is, to be honest.

File: includes/MediaWiki/ReconciliationItem.php [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+namespace MediaWiki\Extension\WikibaseReconcileEdit\MediaWiki;
+
+use Wikibase\DataModel\Entity\Item;
+
+/**
+ * Class that contains the item and revision found by the reconciliation service
+ * @see ReconciliationService
+ */
+class ReconciliationItem {
```
Comment (lucaswerkmeister) [inline includes/MediaWiki/ReconciliationItem.php:11]: Maybe we could use the Wikibase `EntityRevision` class instead of this? (Though in that case the content would only be known to be an entity, not specifically an item.)

File: includes/MediaWiki/ReconciliationService.php [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+	/**
+	 * Finds existing item or returns a new one if no matching item is found.
+	 * Lookup is matching on items with statement ($reconcileUrlProperty) set to the $reconciliationUrl
+	 *
+	 * @param PropertyId $reconcileUrlProperty
+	 * @param string $reconciliationUrl
+	 * @return ReconciliationItem
+	 * @throws \Exception
+	 */
+	public function getItemByStatementUrl(
```
Comment (lucaswerkmeister) [inline includes/MediaWiki/ReconciliationService.php]: What do you think about changing “get” to “getOrCreate” here, to make it clearer that the method may create a new item?

File: tests/unit/MediaWiki/ReconciliationServiceTest.php (hunk 4 above)
Comment (toban) [reply tests/unit/MediaWiki/ReconciliationServiceTest.php]: grand, lets remove it

File: includes/MediaWiki/ReconciliationService.php (hunk 6 above)
Comment (toban) [reply includes/MediaWiki/ReconciliationService.php]: yeah sounds good

File: includes/ServiceWiring.php (hunk 1 above)
Comment (toban) [reply includes/ServiceWiring.php]: done

File: includes/ServiceWiring.php (hunk 2 above)
Comment (toban) [reply includes/ServiceWiring.php:38]: done

File: includes/MediaWiki/ReconciliationService.php [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+	}
+
+	/**
+	 * @param int[] $pageIds
+	 * @return ItemId[]
+	 */
+	private function getItemIdsFromPageIds( array $pageIds ) : array {
+		$titles = $this->titleFactory->newFromIDs( $pageIds );
+
+		if ( empty( $titles ) ) {
```
Comment (lucaswerkmeister) [inline includes/MediaWiki/ReconciliationService.php]: Is this added condition necessary? Do you know if anything breaks if we call `EntityIdLookup::getEntityIds()` with an empty array?

File: includes/MediaWiki/ReconciliationService.php [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+
+		// If we have more than one item matches, something is wrong and we can't edit
+		if ( count( $itemsThatReferenceTheUrlInCorrectStatement ) > 1 ) {
+			die( 'Matched multiple Items during reconciliation :(' );
+		}
+
+		// Get our base
+		if ( count( $itemsThatReferenceTheUrlInCorrectStatement ) === 1 ) {
+			$base = $itemsThatReferenceTheUrlInCorrectStatement[0];
+			// XXX: This bit is so annoying...
```
Comment (lucaswerkmeister) [inline includes/MediaWiki/ReconciliationService.php:116]: We can probably alleviate this by using an `EntityRevisionLookup` rather than an `EntityLookup` in the earlier block where we load the item, though that should be done in a separate commit.

File: includes/MediaWiki/ReconciliationItem.php (hunk 5 above)
Comment (toban) [reply includes/MediaWiki/ReconciliationItem.php:11]: Which would mean we would have to keep checking for type and cast it (or refactor all other things relying on this being an item). 

Do we see a use-case for anything other than items? I guess we could extend EntityRevision with some helper function that casts the item but that feels kinda awkward too considering how simple both classes are.

Comment (toban) [reply includes/MediaWiki/ReconciliationItem.php:11]:
```

/**
 * Class that contains the item and revision found by the reconciliation service
 * @see ReconciliationService
 */
class ReconciliationItem extends EntityRevision {

	public function __construct( Item $item, $revision ) {
		parent::__construct( $item, $revision );
	}

	/**
	 * @return Item
	 */
	public function getItem(): Item {
		/** @var Item $item */
		$item = $this->getEntity();
		return $item;
	}
}
```
¯\_(ツ)_/¯

File: includes/MediaWiki/ReconciliationService.php (hunk 7 above)
Comment (toban) [reply includes/MediaWiki/ReconciliationService.php]: If not then I had to add a bunch of mocks for the functions these services use in the test. Which just makes the tests harder to read and more work than exiting early here which should be fine.

File: includes/MediaWiki/ReconciliationItem.php (hunk 5 above)
Comment (lucaswerkmeister) [reply includes/MediaWiki/ReconciliationItem.php:11]: We could even just do

```php
/** @var Item $item */
$item = $entityRevision->getEntity();
```

and call it a day, since PHP is a dynamic language…

Comment (toban) [reply includes/MediaWiki/ReconciliationItem.php:11]: After a chat, lets keep the thing as it is and maybe refactor this later.

File: includes/MediaWiki/ReconciliationService.php (hunk 7 above)
Comment (toban) [reply includes/MediaWiki/ReconciliationService.php]: lets get rid of it

File: includes/MediaWiki/ReconciliationService.php (hunk 8 above)
Comment (toban) [reply includes/MediaWiki/ReconciliationService.php:116]: put this in it's own task https://phabricator.wikimedia.org/T282553 so we don't forget
